coderd/agentapi/audit.go

```diff
@@ -0,0 +1,117 @@
+
+	"cdr.dev/slog"
+
+	agentproto "github.com/coder/coder/v2/agent/proto"
+	"github.com/coder/coder/v2/coderd/audit"
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/codersdk/agentsdk"
+)
+
+type AuditAPI struct {
+	AgentFn  func(context.Context) (database.WorkspaceAgent, error)
+	Auditor  *atomic.Pointer[audit.Auditor]
+	Database database.Store
+	Log      slog.Logger
+}
+
+func (a *AuditAPI) ReportConnection(ctx context.Context, req *agentproto.ReportConnectionRequest) (*emptypb.Empty, error) {
+	// We will use connection ID as request ID, typically this is the
+	// SSH session ID as reported by the agent.
+	connectionID, err := uuid.FromBytes(req.GetConnection().GetId())
+	if err != nil {
+		return nil, xerrors.Errorf("connection id from bytes: %w", err)
+	}
+
+	var action database.AuditAction
+	switch req.GetConnection().GetAction() {
+	case agentproto.Connection_CONNECT:
+		action = database.AuditActionConnect
+	case agentproto.Connection_DISCONNECT:
+		action = database.AuditActionDisconnect
+	default:
+		return nil, xerrors.Errorf("unknown agent connection action %q", req.GetConnection().GetAction())
+	}
+
+	var connectionType agentsdk.ConnectionType
+	switch req.GetConnection().GetType() {
+	case agentproto.Connection_SSH:
+		connectionType = agentsdk.ConnectionTypeSSH
+	case agentproto.Connection_VSCODE:
+		connectionType = agentsdk.ConnectionTypeVSCode
+	case agentproto.Connection_JETBRAINS:
+		connectionType = agentsdk.ConnectionTypeJetBrains
+	case agentproto.Connection_RECONNECTING_PTY:
+		connectionType = agentsdk.ConnectionTypeReconnectingPTY
+	default:
+		return nil, xerrors.Errorf("unknown agent connection type %q", req.GetConnection().GetType())
+	}
+
+	// Fetch contextual data for this audit event.
+	workspaceAgent, err := a.AgentFn(ctx)
+	if err != nil {
+		return nil, xerrors.Errorf("get agent: %w", err)
+	}
+	workspace, err := a.Database.GetWorkspaceByAgentID(ctx, workspaceAgent.ID)
+	if err != nil {
+		return nil, xerrors.Errorf("get workspace by agent id: %w", err)
+	}
+	build, err := a.Database.GetLatestWorkspaceBuildByWorkspaceID(ctx, workspace.ID)
+	if err != nil {
+		return nil, xerrors.Errorf("get latest workspace build by workspace id: %w", err)
+	}
+
+	// We pass the below information to the Auditor so that it
+	// can form a friendly string for the user to view in the UI.
+	type additionalFields struct {
+		audit.AdditionalFields
+
+		ConnectionType agentsdk.ConnectionType `json:"connection_type"`
+		Reason         string                  `json:"reason,omitempty"`
+	}
+	resourceInfo := additionalFields{
+		AdditionalFields: audit.AdditionalFields{
+			WorkspaceID:    workspace.ID,
+			WorkspaceName:  workspace.Name,
+			WorkspaceOwner: workspace.OwnerUsername,
+			BuildNumber:    strconv.FormatInt(int64(build.BuildNumber), 10),
+			BuildReason:    database.BuildReason(string(build.Reason)),
+		},
+		ConnectionType: connectionType,
+		Reason:         req.GetConnection().GetReason(),
```

Comment:
Would it make sense to have some well-defined reasons here? Or is this basically going to be a place to stuff an `err.Error()`?

Reply:
What would be the use-case? We're planning to use it as free-text to add some more context as to why a user disconnected. Perhaps process exited successfully / failed, or file transfer blocked, etc. It'll be similar to what's already logged, I'm thinking. For now I think free-text is good enough.

Reply:
That makes sense!